### PR TITLE
feat: add match note schema and file creation services

### LINKED ISSUE #13: Define markdown/frontmatter schema for match notes
## Summary
Define the match-note structure before implementation so generated notes stay stable, readable, and query-friendly.

## Why
The first MVP workflow depends on a predictable note format. If the schema is left vague, later implementation work may hard-code inconsistent fields, create migration pain, or make Dataview usage awkward.

This issue should establish the canonical shape that later note-generation code follows.

## Scope
- Define the frontmatter fields for match notes.
- Define the section layout for generated notes.
- Decide which sections are machine-managed versus user-authored.
- Keep Dataview compatibility in mind.
- Decide where the canonical schema should live:
  - documented in Markdown, and/or
  - represented in code via a template builder or schema-oriented service.

## Proposed schema direction
At minimum, the schema should cover fields such as:
- `type`
- `sport`
- `source_url`
- `source_host`
- `competition`
- `match_date`
- `home_team`
- `away_team`
- tags

The body should likely include sections such as:
- Snapshot
- Lineups
- Match stats
- Timeline
- My observations
- Tactical notes

## Questions to resolve
- Which fields are required versus optional at MVP time?
- Which sections are safe for the plugin to manage automatically?
- Which sections must remain user-authored and never be overwritten later?
- What filename/title conventions should the schema support?

## Definition of done
- The schema is documented and stable enough for implementation.
- There is one clearly defined canonical structure for match notes.
- Follow-up implementation issues can reference this issue as the source of truth.

### LINKED ISSUE #18: Implement match note file creation and templating
## Summary
Build the note-generation layer that creates a structured Markdown match note in the configured vault folder.

## Why
The command workflow needs a real file-generation step. Without it, URL input and schema design are disconnected from the actual MVP behavior.

This implementation should follow the schema defined in the match-note schema issue rather than inventing its own independent note structure.

## Scope
- Generate note filenames.
- Create notes in the configured match-notes folder.
- Render frontmatter and body sections from the agreed schema.
- Handle duplicate filenames or path conflicts sensibly.
- Open the created note or otherwise surface a useful success path.

## Dependencies
This issue should follow the schema defined in:
- `Define markdown/frontmatter schema for match notes`

## Implementation expectations
- Use Obsidian vault APIs for file creation.
- Keep note generation separate from the command entry point.
- Keep templating logic readable and easy to revise.
- Avoid mixing provider-specific fetching into this layer.

## Definition of done
- Running the feature flow creates a readable, structured Markdown note in the configured vault location.
- The generated note follows the schema agreed in the schema issue.
- Duplicate handling fails safely and predictably.

### PR DESCRIPTION
**Pull request by FerdiHS**
This pull request introduces the initial implementation for generating football match notes, including a canonical schema, file creation logic, and consistent template structure. The changes establish a predictable and extensible format for match notes and ensure safe file creation within the Obsidian vault.

**Match note schema and template standardization:**

* Added `docs/match-note-schema.md` to define the canonical MVP schema for football match notes, specifying required frontmatter fields, reserved fields for future use, section ordering, and ownership/naming conventions.
* Updated the match note template in `src/services/match-note-template.ts` to use the new schema: required frontmatter fields (`type`, `sport`, `source_url`), and a standardized set of section headings with no starter prose. [[1]](diffhunk://#diff-d4d9b084911ae5ccb42abd4e7adade8cd4d09e5e6c2babd0b991b45cc613c942R4-R18) [[2]](diffhunk://#diff-d4d9b084911ae5ccb42abd4e7adade8cd4d09e5e6c2babd0b991b45cc613c942R31-R42)

**File creation and path management:**

* Added `src/services/match-note-files.ts` to handle creation of match note files in the vault, including logic to ensure folders exist and to avoid filename collisions by incrementing a numeric suffix.
* Added `src/services/match-note-paths.ts` to generate normalized file paths, timestamped filenames, and folder creation chains for match notes.

**Related Issues:**

* Resolve #13 
* Resolve #18

### CODE REVIEW

**Comment by FerdiHS** [comment]
@codex, please review this PR with a focus on:
- correctness of the match note schema
- filename/path generation and duplicate handling
- vault file creation behavior
- regressions or missing edge cases
- missing tests or validation gaps

File: src/services/match-note-files.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+export async function createMatchNoteFile(vault: Vault, input: MatchNoteInput): Promise<TFile> {
+	const draft = createMatchNoteDraft(input);
+	const now = new Date();
+
+	await ensureFolderExists(vault, draft.folder);
+
+	const filePath = findAvailableMatchNotePath(vault, draft.folder, draft.title, now);
+
+	return await vault.create(filePath, draft.content);
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-note-files.ts]: Thanks for spotting. I have implemented the fix in the next commit.

The file-creation flow now retries after create-time conflicts, so concurrent note creation can advance to `... 2.md`, `... 3.md`, etc. instead of failing. Folder creation was also made race-tolerant.

**Comment by FerdiHS** [comment]
@codex review

Please review this PR again with a focus on:
- match note schema and template alignment
- note file creation behavior in `src/services/match-note-files.ts`
- duplicate filename collision handling, including concurrent creation edge cases
- folder creation behavior
- regressions, typos, and missing validation gaps

File: src/services/match-note-template.ts [hunk 2]
```diff
@@ -13,25 +28,18 @@ export function createMatchNoteDraft(input: MatchNoteInput): MatchNoteDraft {
 }
 
 function buildMatchNoteMarkdown(sourceUrl: string): string {
+	const sectionHeadings = MATCH_NOTE_SECTIONS.flatMap((section) => [`## ${section}`, '']);
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-note-template.ts]: Thanks for noticing. I replaced the `flatMap` section builder with a simple loop so the template no longer depends on `Array.prototype.flatMap` and stays aligned with the repo’s current TypeScript lib surface.

**Comment by FerdiHS** [comment]
@codex review

Please review this PR again with a focus on:
- match note schema and template alignment
- note file creation behavior
- filename/path generation and duplicate handling
- folder creation behavior
- regressions, typos, and missing validation gaps

File: src/services/match-note-paths.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import { normalizePath } from 'obsidian';
+
+const MATCH_NOTE_FILE_EXTENSION = '.md';
+
+export function createMatchNoteTimestamp(now: Date): string {
+	return [
+		now.getFullYear().toString().padStart(4, '0'),
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-note-paths.ts]: I replaced `padStart` in the timestamp helper with a small local padding helper so the code no longer depends on newer string APIs and stays aligned with the repo’s current TypeScript lib surface.

**Comment by FerdiHS** [comment]
@codex review

Please review this PR again with a focus on:
- TypeScript/lib compatibility of the changed helpers
- match note schema and template alignment
- filename/path generation and duplicate handling
- folder normalization and vault file creation behavior
- regressions, edge cases, and missing validation gaps

File: src/services/match-note-template.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-note-template.ts]: I removed the helper usage outside the repo’s configured TypeScript lib surface:
- replaced `flatMap` with a simple loop
- replaced `padStart` with a small local `padNumber(...)` helper

**Comment by FerdiHS** [comment]
@codex review

Please review the latest pushed changes again with a focus on:
- TypeScript/lib compatibility of the changed helpers
- match note schema and template alignment
- filename/path generation and duplicate handling
- folder normalization and vault file creation behavior
- regressions, edge cases, and missing validation gaps

**Comment by FerdiHS** [comment]
@codex, please review this PR with a focus on:
- correctness of the match note schema
- filename/path generation and duplicate handling
- vault file creation behavior
- regressions or missing edge cases
- missing tests or validation gaps

Make sure that this PR is safe to be merged.

File: src/types.ts [hunk 4]
```diff
@@ -1,9 +1,19 @@
...
 
 export function normalizeMatchNotesFolder(folder: string): string {
 	const trimmedFolder = folder.trim();
 
-	return trimmedFolder.length > 0 ? trimmedFolder : DEFAULT_MATCH_NOTES_FOLDER;
+	if (trimmedFolder.length === 0) {
+		return DEFAULT_MATCH_NOTES_FOLDER;
+	}
+
+	const normalizedFolder = normalizePath(trimmedFolder).replace(/^\/+/, '');
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/types.ts:16]: Fixed in the latest changes. `normalizeMatchNotesFolder()` now rejects parent-directory segments and falls back to the default notes folder, and the path helpers sanitize again before folder/file creation, so the configured path can’t escape the vault.

**Comment by FerdiHS** [comment]
@codex review

Please review the latest changes again, with a focus on:
- validation and path handling
- file creation and duplicate handling
- regressions or remaining edge cases

**Comment by FerdiHS** [comment]
@codex review

Please do one final pass and confirm whether this PR is ready and safe to merge.

Focus on:
- correctness
- regressions and edge cases
- path and file-creation safety
- any remaining validation gaps

If there are no issues, please confirm that the PR looks safe to merge.
